### ISSUE TITLE
Wrong number of type arguments handling

### ISSUE THREAD

**Issue by scheglov** [dart-lang/sdk#31062]
```dart
class A<K, V> {
  K element;
}
main(A<int> a) {
  a.element.anyGetterExistsInDynamic;
}
```

```kernel
library;
import self as self;
import "dart:core" as core;

class A<K extends core::Object, V extends core::Object> extends core::Object {
  field self::A::K element = null;
  default constructor •() → void
    : super core::Object::•()
    ;
}
static method main(self::A<core::int, dynamic> a) → dynamic {
  a.element.anyGetterExistsInDynamic;
}
```

Is this correct?
`19.8 Parameterized Types` says that all type arguments should be `dynamic`.

Paul, is this a type inference issue?

**Comment by stereotype441** [comment]
Looks like the culprit is in `pkg/front_end/lib/src/fasta/kernel/kernel_builder.dart`:

```dart
List<DartType> computeDefaultTypeArguments(LibraryBuilder library,
    List<TypeParameter> typeParameters, List<DartType> arguments) {
  // TODO(ahe): Not sure what to do if `arguments.length !=
  // cls.typeParameters.length`.
  if (arguments == null) {
    return new List<DartType>.filled(
        typeParameters.length, const DynamicType());
  }
  if (arguments.length < typeParameters.length) {
    arguments = new List<DartType>.from(arguments);
    for (int i = arguments.length; i < typeParameters.length; i++) {
      arguments.add(const DynamicType());
    }
  } else if (arguments.length > typeParameters.length) {
    return arguments.sublist(0, typeParameters.length);
  }
  return arguments;
}
```

Reassigning to Peter since he's more familiar with this code than I am.

**Comment by sigmundch** [comment]
I believe it's old recovery logic that we want to change (Peter can confirm).

When I've discussed this with Peter in the past, I recall we agreed to change to ignore the provided arguments and treat the type as if it was written as a raw type (as in `main(A a)`). The idea of padding with dynamic was that we could use as much of the data from the user, but in reality this can yield surprising behavior if we pad in the wrong place (maybe they wrote what was supposed to be the second argument, not the first one?)

In the new logic I added for type declarations (e.g. on top-level fields like `A<int> a;`) we have adopted this new approach.
